daemon/src/lib.rs

```diff
@@ -0,0 +1,61 @@
+// Copyright © 2019-2020 The Radicle Foundation <hello@radicle.foundation>
+//
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+//! Abstractions and utilities to run and interact with link and surf.
+
+#![warn(
+    clippy::all,
+    clippy::cargo,
+    clippy::nursery,
+    clippy::pedantic,
+    clippy::unwrap_used,
+    missing_docs,
+    unused_import_braces,
+    unused_qualifications
+)]
+#![allow(
+    clippy::cargo_common_metadata,
+    clippy::clone_on_ref_ptr,
+    clippy::expect_used,
+    clippy::implicit_return,
+    clippy::integer_arithmetic,
+    clippy::missing_inline_in_public_items,
+    clippy::multiple_crate_versions,
+    clippy::multiple_inherent_impl,
+    clippy::similar_names,
+    clippy::too_many_lines
+)]
+#![feature(duration_zero, hash_set_entry, never_type)]
+
+#[cfg(test)]
+#[macro_use]
+extern crate assert_matches;
+
+pub use librad::{
```

Comment:
Is this still necessary given that downstreams have to depend on the repo anyway?

Reply:
They don't need to depend on `librad`, they have to depend on `radicle-git-ext`. I think it would make sense just to re-export `librad` though and not specify the other re-exports. Whatya think?

Reply:
Originally the reexport of `librad` was suppose to hide away the link traces from the api crate which dependent on coco and coco on librad. As we now know that was a bad idea to begin with and downstream has to get a repo dependency anyway, if it need librad types it should depend on it directly imho. It also appears odd that here a re-export of another crate from the same workspace takes place.

Reply:
> As we now know that was a bad idea to begin with and downstream has to get a repo dependency anyway

But it doesn't need a `librad` repo dependency is what I'm saying, the sum total of deps it needs to declare could be `radicle-daemon` and `radicle-git-ext`. The latter being an unfortunate circumstance of `radicle-macros` functions expanding to `::radicle_git_ext`.

> It also appears odd that here a re-export of another crate from the same workspace takes place.

This also happens in `librad` with `macros`, `git-ext`, and `std-ext`.

So the issue with having multiple upstream declarations is you end up in this situation:
```
[dependency.foo]
git = "<dev/foo>"
rev = "abc"

[dependency.bar]
git = "<dev/bar>"
rev = "abc"
```

and now every time they update `"abc"` they have to do it everywhere. If they don't then they get weird-ass errors where the types are the same but come from different versions of the project -- we saw this with `radicle-git-ext` when fixing up some upstream code.

Re-exporting `librad` from `radicle-daemon` will mean that the types used are always on the same versioning, which is very useful IMO.

The alternatives to re-exporting are:
1. Require downstream users to add deps with the SHAs that they have to manage manually
2. Require downstream users to point to a particular branch, i.e. `master`, for all related deps and run `cargo update` to get the latest changes 

So ya, we could let them add deps in these manners :) I suppose we just have to weigh up the options (or maybe there's options I'm missing?). @geigerzaehler might be interested in this as well :eyes:

Reply:
> Re-exporting `librad` from `radicle-daemon` will mean that the types used are always on the same versioning, which is very useful IMO.

I agree and would prefer to have `librad` re-exported from `radicle-deamon`.

> I think it would make sense just to re-export `librad` though and not specify the other re-exports. Whatya think?

That would be a good idea. The API surface is already quite large and complicated so anything we can do to streamline this would be appreciated.

Reply:
Please keep in mind that a daemon is a process, not a library. As such, the surface API of this crate should be equivalent to its HTTP API.

Some kind of "prelude" crate which exports things to build your own daemon (or whatever application) is a separate concern. I am a strong proponent of designing libraries in a way such that even "internal" / unstable calls are accessible to consumers, which is not something Rust crates are very good at, and one of the reasons I'm in favour of a crate-zoo which can be recomposed into whatever preferences users might have.

Reply:
Thanks for the input @geigerzaehler and @kim :bow: 

> Please keep in mind that a daemon is a process, not a library. As such, the surface API of this crate should be equivalent to its HTTP API.

So because I wanted to stage these moves there is no HTTP API in here yet. This means that we need to have some parity for re-exporting until we move that portion of the code over. 

With that in mind, I would say we keep this as is to keep parity -- which is currently green https://github.com/radicle-dev/radicle-upstream/pull/1753. Then when we move on to the next stage of moving things over we can remove the re-exports here and the HTTP API can rely on `path = ../librad`. 

Does that sound reasonable to yis?